### ISSUE TITLE
508-defect-1 ❗ Launchblocker [SEMANTIC MARKUP, COGNITION]: Groups of form elements MUST have group labels (e.g. <fieldset> and <legend>)

### ISSUE THREAD

**Issue by jenstrickland** [department-of-veterans-affairs/va.gov-team#16224]
# [508-defect-1 :exclamation: Launchblocker](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/accessibility/guidance/defect-severity-rubric.md#508-defect-1)

## Feedback framework

- **❗️ Must** for if the feedback must be applied 
- **⚠️ Should** if the feedback is best practice 
- **✔️ Consider** for suggestions/enhancements 

## Definition of done

1. Review and acknowledge feedback.
1. Fix and/or document decisions made.
1. Accessibility specialist will close ticket after reviewing documented decisions / validating fix.

<hr/>

## Point of Contact

**VFS Point of Contact:** Josh

## User Story or Problem Statement

As a screen reader user, I need to know that a range of form inputs are grouped together, so that I may understand the scope of the use case.

## Details

Groups of form elements **MUST** have group labels (e.g. `<fieldset>` and `<legend>`, to meet Section 508 requirements. The current code wraps "Select the issue you would like to have reviewed(*Required)" in a `<strong>` element, rather than the `<legend>` which would convey that the group of items within the `<fieldset>` are semantically connected. 

## Acceptance Criteria

- [ ] Change the html as follows, or an alternative standards-compliant manner that programmatically associates the labelling content with the inputs

### Current code

```html
<strong>Select the issue you would like to have reviewed</strong><span class="schemaform-required-span vads-u-font-weight--normal vads-u-font-size--base">(*Required)</span>`
```

### Recommended code

```html
<legend>Select the issue you would like to have reviewed <span class="schemaform-required-span vads-u-font-weight--normal vads-u-font-size--base">(*Required)</span></legend>
```

## Environment

* Operating System: all
* Browser: all
* Screenreading device: all
* Server destination: staging

## Steps to Recreate

1. Enter `https://staging.va.gov/decision-reviews/higher-level-review/request-higher-level-review-form-20-0996/contested-issues` in browser
2. Navigate to Step 2 of 4: Contested issues
3. Start screen reader, and read through the form area
4. Verify that there isn't a semantic correlation between "Select the issue you would like to have reviewed (*Required)" and the following form options

### Current

```diff
<form class="rjsf" novalidate="">
  <fieldset>
    <div class="input-section schemaform-field-container">
      <div class="schemaform-block-header">
!       <strong>Select the issue you would like to have reviewed</strong><span class="schemaform-required-span vads-u-font-weight--normal vads-u-font-size--base">(*Required)</span>
      </div>
      <div>
        <div class="schemaform-field-template usa-input-error schemaform-first-field" tabindex="-1">
          <label 
```

### Recommendation

```html
<form class="rjsf" novalidate="">
  <fieldset>
    <div class="input-section schemaform-field-container">
      <div class="schemaform-block-header">
        <legend>Select the issue you would like to have reviewed <span class="schemaform-required-span vads-u-font-weight--normal vads-u-font-size--base">(*Required)</legend>
      </div>
      <div>
        <div class="schemaform-field-template usa-input-error schemaform-first-field" tabindex="-1">
          <label 
```

### New Recommendation

The legend **must** be nested directly under the fieldset for it to be accessibly read out to screen reader users when interacting with the checkboxes below.

```diff
<form class="rjsf" novalidate="">
  <fieldset>
+   <legend>Select the issue you would like to have reviewed <span class="schemaform-required-span vads-u-font-weight--normal vads-u-font-size--base">(*Required)</legend>
    <div class="input-section schemaform-field-container">
-      <div class="schemaform-block-header">
-         <legend>Select the issue you would like to have reviewed <span class="schemaform-required-span vads-u-font-weight--normal vads-u-font-size--base">(*Required)</legend>
-      </div>
      <div>
        <div class="schemaform-field-template usa-input-error schemaform-first-field" tabindex="-1">
          <label 
```

## WCAG or Vendor Guidance (optional)

* [WCAG 2.0 Level A - 1.3.1 - Info and Relationships: Information, structure, and relationships conveyed through presentation can be programmatically determined or are available in text.](https://www.w3.org/TR/UNDERSTANDING-WCAG20/content-structure-separation-programmatic.html)

* [Wuhcag 1.3.1 – Info and Relationships (Level A)](https://www.wuhcag.com/info-and-relationships/)

* [WCAG 2.0 Level A - 4.1.1 - Parsing](http://www.w3.org/TR/UNDERSTANDING-WCAG20/ensure-compat-parses.html)

## Screenshots or Trace Logs

![Screen Shot 2020-11-16 at 2 35 28 PM](https://user-images.githubusercontent.com/57469/99299425-339aab80-2819-11eb-9c43-1896338c55c2.png)

### New Recommendation
<img width="1077" alt="Screen Shot 2021-01-07 at 10 00 18 AM" src="https://user-images.githubusercontent.com/14154792/103909303-8a15bf00-50d1-11eb-858f-ff341e794b6d.png">

**Comment by Mottie** [comment]
I believe this has already been implemented... I'll verify once I have a second to check.

**Comment by Mottie** [comment]
Verified that this has been fixed.

**Comment by joshkimux** [comment]
@kevinstachura @Mottie @luke-gcio This page will most likely require a meeting for us to talk through as it's a bit complicated. Topics would include:

- Splitting up the page itself into two separate fieldsets OR moving the "same office" checkbox to its own separate page
- How to handle screen reader experience for the "same office" checkbox
- Ways to move forward with the error alert in consideration of design/form system limitations

**Comment by joshkimux** [comment]
Some more context for that discussion below:

<img width="1792" alt="Screen Shot 2020-12-19 at 4 46 17 PM" src="https://user-images.githubusercontent.com/14154792/102700303-44747d80-421a-11eb-9d1e-2af128bdfb9a.png">

<img width="1075" alt="Screen Shot 2020-12-19 at 4 48 15 PM" src="https://user-images.githubusercontent.com/14154792/102700304-48080480-421a-11eb-89d8-d7955dc3e27a.png">

**Comment by joshkimux** [comment]
I need to dig around a bit more as Safari/Voiceover announces the error + "and one more item, group" which may be related to this conversation.

<img width="782" alt="Screen Shot 2020-12-20 at 4 40 28 PM" src="https://user-images.githubusercontent.com/14154792/102725187-ccbe5580-42e2-11eb-9dbc-8fe57de86bc1.png">

**Comment by joshkimux** [comment]
@Mottie , just tested the updated page with NVDA/firefox and realized that the original recommendation on this ticket needs to be updated. [Legends must be directly nested under the fieldset in order to work.](https://www.accessibility-developer-guide.com/examples/forms/grouping-with-fieldset-legend/). I've attached a screenshot below of a click drag/drog in the inspector that fixed the issue, but I'm unsure if the form system will allow for this.

<img width="1077" alt="Screen Shot 2021-01-07 at 10 00 18 AM" src="https://user-images.githubusercontent.com/14154792/103907834-c811e380-50cf-11eb-9b98-5891f27f26c3.png">

**Comment by joshkimux** [comment]
@Mottie noted this update may need to be addressed with the form system team. @caw310 I leave this in your capable hands 🙇

**Comment by Mottie** [comment]
Just to be clear, we're unable to make the `<legend>` a direct child of the `<fieldset>` using the form system. The accessibility tools don't call this problem out as long as there is a `<legend>` within the `<fieldset>`, but [per the legend element spec](https://html.spec.whatwg.org/multipage/form-elements.html#the-legend-element), the `legend` must be "the first child of a fieldset element.".

**Comment by joshkimux** [comment]
@Mottie I went through staging again this morning and caught one more item to be addressed just in case it wasn't on your radar already! Would we be able to change "Your Availability" into a legend worded "Please indicate your availability"? This will make it consistent with the changes we have in place for representative's information. (see screenshots below)

<img width="1417" alt="Screen Shot 2021-01-11 at 10 13 31 AM" src="https://user-images.githubusercontent.com/14154792/104200186-5bac2280-53f6-11eb-90f1-1fc47fe26746.png">

<img width="1401" alt="Screen Shot 2021-01-11 at 10 15 23 AM" src="https://user-images.githubusercontent.com/14154792/104200024-299ac080-53f6-11eb-8a9d-53610c98384f.png">

**Comment by Mottie** [comment]
Thanks Josh! This fix should be available in staging soon!

**Comment by joshkimux** [comment]
Confirmed this was fixed! Closing out!
